restic check with snapshot filters (continuation of PR #5213)

wplapper opened the pull request:
This is the continuation of PR #5213. I decided to use a git provided e-mail address instead of my private e-mail address. All changes up-to 2025-03-07 have been incorporated. Details to follow here:

changelog/unreleased/issue-3326:
Introduction of snapshot based filter for command `restic check`

cmd/restic/cmd_check.go:
Integration of snapshot based filtering. Some messages had to be changed so that the text does not refer to 'all packfiles' but to 'selected packfiles'.

cmd/restic/cmd_check_integration_test.go:
A new test for filtering via snapshots has been added.

doc/077_troubleshooting.rst:
A short paragraph in the manual has been added about snapshot based filtering.

internal/checker/checker.go:
A new func CheckWithSnapshots() has been added which collects the needed packfiles via restic.FindUsedBlobs().

internal/checker/checker_test.go:
A new test TestCheckRepoSnapshot() has been added to check out the new functionality.

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
This PR addresses snapshot based filters in command ``restic check``.
<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
https://github.com/restic/restic/issues/3326
<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
https://github.com/restic/restic/issues/3326

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [X] I have added tests for all code changes.
- [X] I have added documentation for relevant changes (in the manual).
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [X] I'm done! This pull request is ready for review.

MichaelEischer reviewed (changes requested):
The changes from https://github.com/restic/restic/pull/5532 require untangling how the pack file selection works here, see the comments for details. `Structure(...)` should also adhere to the snapshot selection. That method can also generate the pack file list, making `CheckWithSnapshots` unnecessary and also avoiding the massive performance penalty of iterating a second time over the snapshot trees.

Edit:
Just integrate the snapshot selection into `checker.LoadSnapshots`. Then `checker.Snapshots` naturally also adheres to the snapshot filter. The `checker.Checker` can the also wrap the `ReadPacks` method such that it replaces the pack list passed to the filter callback with the one only containing the packs relevant for the snapshots.

wplapper commented:
Hi [MichaelEischer](https://github.com/MichaelEischer),

looking at the existing changes after #5532 and the proposed changed in #5550 I have decided to wait until #5550 has been commited. Then I will start from a ``restic 0.18.1`` base and adopt my changes and your recommendations accordingly. I will place my PR into "draft" status.

wplapper committed:
restic check with snapshot filter:

New implementation based on PR#5550 (refactor pack selection for read data):

cmd/restic/cmd_check.go: had to move the filter creation code buildPacksFilter()
to the point where the repo is open and the index has been loaded. Otherwise
the new function selectPacks() would not work.

Added the new function selectPack() which creates a packfile list based on current
snapshot filters. selectPack() creates a list of used snapshots, uses
data.FindUsedBlobs() to create a set of used blobs and uses this set to generate a map
of used packfiles.

In buildPacksFilter() code has been inserted for the print output to indicate
snapshot filter selection.

doc/077_troubleshooting.rst adds a paragraph about snapshot filtering.

changelog/unreleased/issue-3326 announces the new enhancement.

wplapper committed:
restic check - snapshots - linter error

linter caught me out!

wplapper committed:
restic check snapshot filter: code rework

cmd/restic/cmd_check.go:
moved filter checking into internal/checker/checker.go
as a consequence, checker.LoadSnapshots() gained thwo paramenters,
which were just saved away. The actual processing of the snapshot filter got moved
into checker.Structure(), where context and repository are available.
function buildPacksFilter() gained the parameter 'chkr', so the selected packs
and the active flag can be retrieved from the checker instance.
buildPacksFilter() has been modified, so there is a choice between 'normal'
processing and snapshot filter processing. The printed messages were adopted
accordingly.

cmd/restic/cmd_check_integration_test.go: three new tests verifying the filtering.

internal/checker/checker.go: seee above

internal/checker/checker.go and internal/checker/testing.go had to be modified for
using the changed LoadSnapshots() interface.

wplapper committed:
restic check with snapshot filter - linter check

linter caught me out with an incorrect spelling

wplapper committed:
restic check with filtered snapshot

add subtest in cmd_check_integration_test.go:
show that exaxctly one snapshot has been selected in Structure()

wplapper committed:
restic check with snapshot filter - windows chokes

I have no idea whu Windows chokes in a different place from here. Timing?

wplapper committed:
restic check - reworked ReadPacks() wrapper

cmd/restic/cmd_check.go:
moved a lot of code from the command to internal/checker/checker
inserted one message line "Snapshot filtering is active"

internal/checker/checker.go:
split the code from cmd/restic/cmd_check.go into parts for LoadSnapshots(),
Structure() and a new wrapper for ReadPacks()

cmd/restic/cmd_check_integration_test.go:
adopted the tests to the new environment

wplapper committed:
restic check - fix linter issues

linter caught me out again

wplapper committed:
restic check snap filter - more messages on error

cmd/restic/cmd_check.go:
passing printer parameter to chkr.Structure()

cmd/restic/cmd_check_integration_test.go
internal/checker/checker_test.go
internal/checker/testing.go
adopting changes to chker.Structure()

internal/checker/checker.go:
now printing error message to printer.E

wplapper committed:
restic check snap filter error messages - linter

linter caught me out

wplapper committed:
test windows failures

remove these pesky fialing windows tests for the time being

wplapper commented:
I have removed these failing windows tests from the this PR. I will have to wait anyway until this issue has been resolved.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -127,8 +137,32 @@ func loadSnapshotTreeIDs(ctx context.Context, lister restic.Lister, repo restic.
+	}
+
+	if len(trees) == 0 && printer != nil {
+		(*printer).E("\n=== no snapshots selected - expect surprising messages==\n\n")
```
Why not just return in this case? Restic should still behave reasonably, even if a user did something incorrect.

wplapper replied:
done as recommended.

MichaelEischer commented on cmd/restic/cmd_check_integration_test.go:
```diff
@@ -34,3 +35,87 @@ func testRunCheckOutput(t testing.TB, gopts global.Options, checkUnused bool) (s
+func testRunCheckOutputWithOpts(t testing.TB, gopts global.Options, opts CheckOptions, args []string) (string, error) {
+	buf, err := withCaptureStdout(t, gopts, func(ctx context.Context, gopts global.Options) error {
+		gopts.Verbosity = 2
+		gopts.Verbose = 2
```
```suggestion
```
`gopts.Verbose` is unnecessary. Sorry for that mess.

wplapper replied:
removed ``gopts.Verbose = 2``.

MichaelEischer commented on doc/077_troubleshooting.rst:
```diff
@@ -82,6 +82,12 @@ If ``check`` detects damaged pack files, it will show instructions on how to rep
 them using the ``repair pack`` command. Use that command instead of the "Repair the
 index" section in this guide.
 
+If you are interested to check the repository via snapshots, you can now
```
```suggestion
If you are interested to check only specific snapshots, you can now
```

MichaelEischer reviewed (changes requested):
Thanks for the changes so far. This is definitely moving into the right direction, but needs a few further cleanups. The current `ReadPacks` extension is also _far_ slower than it has to be (see comments).

wplapper commented on internal/checker/checker.go:
```diff
@@ -127,8 +137,32 @@ func loadSnapshotTreeIDs(ctx context.Context, lister restic.Lister, repo restic.
+		})
+		if err != nil {
+			if printer != nil {
+				(*printer).E("could not load snapshots: %v", err)
```
I appended the error to `èrrs``. From there it will be taken to ``errChan``.

wplapper commented on internal/checker/checker.go:
```diff
@@ -127,8 +137,32 @@ func loadSnapshotTreeIDs(ctx context.Context, lister restic.Lister, repo restic.
+	var trees []restic.ID
+
+	if len(c.args) > 0 || !c.snapshotFilter.Empty() {
+		err := (&c.snapshotFilter).FindAll(ctx, c.snapshots, c.repo, c.args, func(_ string, sn *data.Snapshot, _ error) error {
```
see above.

wplapper commented on internal/checker/checker.go:
```diff
@@ -127,8 +137,32 @@ func loadSnapshotTreeIDs(ctx context.Context, lister restic.Lister, repo restic.
+	var errs []error
+	var trees []restic.ID
+
+	if len(c.args) > 0 || !c.snapshotFilter.Empty() {
```
done as recpmmended (``loadActiveTrees``).

wplapper commented on internal/checker/checker.go:
```diff
@@ -127,8 +137,32 @@ func loadSnapshotTreeIDs(ctx context.Context, lister restic.Lister, repo restic.
 // been traversed.
-func (c *Checker) Structure(ctx context.Context, p *progress.Counter, errChan chan<- error) {
-	trees, errs := loadSnapshotTreeIDs(ctx, c.snapshots, c.repo)
+func (c *Checker) Structure(ctx context.Context, p *progress.Counter, errChan chan<- error, printer *progress.Printer) {
```
made it a ``progress.Printer``.

wplapper commented on internal/checker/checker.go:
```diff
@@ -259,3 +293,46 @@ func (c *Checker) UnusedBlobs(ctx context.Context) (blobs restic.BlobHandles, er
+	}
+	// convert selected trees to a set of used blobs
+	usedBlobs := restic.NewBlobSet()
+	err = data.FindUsedBlobs(ctx, c.repo, c.trees, usedBlobs, nil)
```
done as recommended.

wplapper commented on internal/checker/checker.go:
```diff
@@ -51,9 +58,12 @@ func New(repo checkerRepository, trackUnused bool) *Checker {
 	var err error
 	c.snapshots, err = restic.MemorizeList(ctx, c.repo, restic.SnapshotFile)
+	c.snapshotFilter = snapshotFilter
+	c.args = args
```
Yep, done as recommended.

wplapper committed:
restic check with SnapshotFilter - implement recommendations

changelog/unreleased/issue-3326:
changed wording

doc/077_troubleshooting.rst:
clarified usage of `restic check` with SnapshotFilter

internal/restorer/restorer_windows_test.go:
reinstated original windows test

cmd/restic/cmd_check.go:
added word 'filtered ' for output display in the various messages

internal/checker/checker.go:
reworked checker logic according to recommendations, removed extra
parameter `printer`in `Structure()`.
added helper function `loadActiveTrees()`
modified `ReadPacks()` to pass filtered packfiles to `repository.ReadPacks()`.

cmd/restic/cmd_check_integration_test.go:
internal/checker/checker_test.go:
internal/checker/testing.go:
adjusted parameter lists for modified functions.

MichaelEischer commented on cmd/restic/cmd_check.go:
```diff
@@ -365,6 +361,12 @@ func runCheck(ctx context.Context, opts CheckOptions, gopts global.Options, args
 		return summary, ctx.Err()
 	}
 
+	opts.filteredStatus = chkr.FilterStatus()
```
Let's move it below the `if opts.CheckUnused {` block. Then the call is directly next to the place where it's used.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -51,9 +57,13 @@ func New(repo checkerRepository, trackUnused bool) *Checker {
 	c.snapshots, err = restic.MemorizeList(ctx, c.repo, restic.SnapshotFile)
+	c.snapshotFilter = snapshotFilter
+	c.args = args
+	c.trackUnused = true
```
That should only be set if the snapshotFilter is actually used. Aka. `c.filterActive` must also be set here. Or even better, get completely rid of `filterActive` and just use a helper function that determines whether there's an active filter.

wplapper replied:
moved setting of ``c.trackUnused`` into ``loadActiveTrees()``.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -124,13 +134,41 @@ func loadSnapshotTreeIDs(ctx context.Context, lister restic.Lister, repo restic.
+			return
+		}
+		c.filterActive = true
+		c.trees = trees
```
Should just return the trees instead of storing them in `c.trees`

MichaelEischer commented on cmd/restic/cmd_check_integration_test.go:
```diff
@@ -34,3 +35,86 @@ func testRunCheckOutput(t testing.TB, gopts global.Options, checkUnused bool) (s
+	return buf.String(), err
+}
+
+func TestCheckFullOutput(t *testing.T) {
```
Please merge the tree tests into a table test (see for example `TestPrepareCheckCache`; you bascially create a struct that contains the expected inputs and output and then use a loop to iterate over all of them). Almost all of their code is shared.

MichaelEischer reviewed (changes requested):
Thanks for keeping up with all the review comments! I have a few further comments, although most of them are nits this time.

wplapper committed:
restic check SnapshotFilter - incorporate changes

cmd/restic/cmd_check.go:
`buildPacksFilter()` received parameter `filteredStatus`, rearranged code

internal/checker/checker.go:
worked on `ReadPacks()`, rearranged code

cmd/restic/cmd_check_integration_test.go:
build a table of tests to run, together with the expeced results

internal/checker/checker_test.go:
internal/checker/testing.go:
adapt code changes in internal/checker/checker.go

MichaelEischer committed:
minor cleanups

MichaelEischer committed:
simplify check if filter is active

MichaelEischer committed:
remove temporary map from pack filtering

MichaelEischer reviewed (approved):
LGTM. Thanks!

I've added a few small cleanups, see the extra commits for details. In particular, only `packsWithSize` is built but not `selectedPacks`. As sets in go are under the hood also maps, we can just directly build the final map. `FilterStatus()` is now the authoritative source whether filtering happens and is less sensitive to whether it's called after certain other methods.

MichaelEischer committed:
fix compatiblity with recent changes

MichaelEischer commented:
And a trivial rebase + fix to work with the changed datatype for `c.blobRefs.M`
